src/graph/util/test/ExpressionUtilsTest.cpp

```diff
@@ -494,6 +494,81 @@ TEST_F(ExpressionUtilsTest, flattenInnerLogicalExpr) {
+  auto listExpr1 = ListExpression::make(pool, elist1);
+
+  auto elist2 = ExpressionList::make(pool);
+  (*elist2).add(ConstantExpression::make(pool, "a")).add(ConstantExpression::make(pool, "b"));
+  auto listExpr2 = ListExpression::make(pool, elist2);
+
+  auto elist3 = ExpressionList::make(pool);
+  (*elist3).add(ConstantExpression::make(pool, 100));
+  auto listExpr3 = ListExpression::make(pool, elist3);
+
+  // a IN [b,c]  ->  a==b OR a==c
+  {
+    auto inExpr1 =
+        RelationalExpression::makeIn(pool, ConstantExpression::make(pool, 10), listExpr1);
+    auto orExpr1 = ExpressionUtils::rewriteInExpr(inExpr1);
+    auto expected1 = LogicalExpression::makeOr(
+        pool,
+        RelationalExpression::makeEQ(
+            pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, 10)),
+        RelationalExpression::makeEQ(
+            pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, 20)));
+    ASSERT_EQ(*expected1, *orExpr1);
+
+    auto inExpr2 =
+        RelationalExpression::makeIn(pool, ConstantExpression::make(pool, "abc"), listExpr2);
+    auto orExpr2 = ExpressionUtils::rewriteInExpr(inExpr2);
+    auto expected2 = LogicalExpression::makeOr(
+        pool,
+        RelationalExpression::makeEQ(
+            pool, ConstantExpression::make(pool, "abc"), ConstantExpression::make(pool, "a")),
+        RelationalExpression::makeEQ(
+            pool, ConstantExpression::make(pool, "abc"), ConstantExpression::make(pool, "b")));
+    ASSERT_EQ(*expected2, *orExpr2);
+  }
+
+  // a IN [b]  ->  a == b
+  {
+    auto inExpr = RelationalExpression::makeIn(pool, ConstantExpression::make(pool, 10), listExpr3);
+    auto expected = RelationalExpression::makeEQ(
+        pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, 100));
+    ASSERT_EQ(*expected, *ExpressionUtils::rewriteInExpr(inExpr));
+  }
+}
+
+TEST_F(ExpressionUtilsTest, rewriteLogicalAndToLogicalOr) {
+  auto orExpr1 = LogicalExpression::makeOr(
+      pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, 20));
+  auto orExpr2 = LogicalExpression::makeOr(
+      pool, ConstantExpression::make(pool, "a"), ConstantExpression::make(pool, "b"));
+
+  // (a OR b) AND (c OR d)  ->  (a AND c) OR (a AND d) OR (b AND c) OR (b AND d)
+  {
+    auto andExpr = LogicalExpression::makeAnd(pool, orExpr1, orExpr2);
+    auto transformedExpr = ExpressionUtils::rewriteLogicalAndToLogicalOr(andExpr);
+
+    std::vector<Expression *> orOperands = {
+        LogicalExpression::makeAnd(
+            pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, "a")),
+        LogicalExpression::makeAnd(
+            pool, ConstantExpression::make(pool, 10), ConstantExpression::make(pool, "b")),
+        LogicalExpression::makeAnd(
+            pool, ConstantExpression::make(pool, 20), ConstantExpression::make(pool, "a")),
+        LogicalExpression::makeAnd(
```

Comment:
Do we support expression as `A and (B or C ) and D` ?

Reply:
`ExpressionUtils::rewriteLogicalAndToLogicalOr` supports the rewrite but this function is only called when the filter contains IN expr for now. We could integrate this into AND expr with OR expr operand later.

Reply:
```
PROFILE lookup on player where  player.name < "zzzzz" and  (player.age in[30, 42]) and  player.name >=  "Tim Duncan"

-----+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------
| id | name      | dependencies | profiling data                                                                                                                   | operator info                                                                                    |
-----+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------
|  3 | Project   | 4            | ver: 0, rows: 2, execTime: 50us, totalTime: 56us                                                                                 | outputVar: [                                                                                     |
|    |           |              |                                                                                                                                  |   {                                                                                              |
|    |           |              |                                                                                                                                  |     "colNames": [                                                                                |
|    |           |              |                                                                                                                                  |       "VertexID"                                                                                 |
|    |           |              |                                                                                                                                  |     ],                                                                                           |
|    |           |              |                                                                                                                                  |     "type": "DATASET",                                                                           |
|    |           |              |                                                                                                                                  |     "name": "__Project_3"                                                                        |
|    |           |              |                                                                                                                                  |   }                                                                                              |
|    |           |              |                                                                                                                                  | ]                                                                                                |
|    |           |              |                                                                                                                                  | inputVar: __Filter_2                                                                             |
|    |           |              |                                                                                                                                  | columns: [                                                                                       |
|    |           |              |                                                                                                                                  |   "COLUMN[0] AS VertexID"                                                                        |
|    |           |              |                                                                                                                                  | ]                                                                                                |
-----+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------
|  4 | IndexScan | 0            | {                                                                                                                                | outputVar: [                                                                                     |
|    |           |              | ver: 0, rows: 2, execTime: 0us, totalTime: 13897us                                                                               |   {                                                                                              |
|    |           |              | storage_detail: {DedupNode:8465(us),IndexOpuputNode:8258(us),IndexScanNode:7456(us),IndexVertexNode:7895(us),RelNode:16610(us),} |     "colNames": [                                                                                |
|    |           |              | "127.0.0.1":39535 exec/total: 8819(us)/13045(us)                                                                                 |       "_vid",                                                                                    |
|    |           |              | }                                                                                                                                |       "player.age",                                                                              |
|    |           |              |                                                                                                                                  |       "player.name"                                                                              |
|    |           |              |                                                                                                                                  |     ],                                                                                           |
|    |           |              |                                                                                                                                  |     "type": "DATASET",                                                                           |
|    |           |              |                                                                                                                                  |     "name": "__Filter_2"                                                                         |
|    |           |              |                                                                                                                                  |   }                                                                                              |
|    |           |              |                                                                                                                                  | ]                                                                                                |
|    |           |              |                                                                                                                                  | inputVar:                                                                                        |
|    |           |              |                                                                                                                                  | space: 1                                                                                         |
|    |           |              |                                                                                                                                  | dedup: false                                                                                     |
|    |           |              |                                                                                                                                  | limit: 9223372036854775807                                                                       |
|    |           |              |                                                                                                                                  | filter:                                                                                          |
|    |           |              |                                                                                                                                  | orderBy: []                                                                                      |
|    |           |              |                                                                                                                                  | schemaId: 2                                                                                      |
|    |           |              |                                                                                                                                  | isEdge: false                                                                                    |
|    |           |              |                                                                                                                                  | returnCols: [                                                                                    |
|    |           |              |                                                                                                                                  |   "_vid",                                                                                        |
|    |           |              |                                                                                                                                  |   "age",                                                                                         |
|    |           |              |                                                                                                                                  |   "name"                                                                                         |
|    |           |              |                                                                                                                                  | ]                                                                                                |
|    |           |              |                                                                                                                                  | indexCtx: [                                                                                      |
|    |           |              |                                                                                                                                  |   {                                                                                              |
|    |           |              |                                                                                                                                  |     "columnHints": [                                                                             |
|    |           |              |                                                                                                                                  |       {                                                                                          |
|    |           |              |                                                                                                                                  |         "endValue": "__EMPTY__",                                                                 |
|    |           |              |                                                                                                                                  |         "beginValue": "30",                                                                      |
|    |           |              |                                                                                                                                  |         "scanType": "PREFIX",                                                                    |
|    |           |              |                                                                                                                                  |         "column": "age"                                                                          |
|    |           |              |                                                                                                                                  |       }                                                                                          |
|    |           |              |                                                                                                                                  |     ],                                                                                           |
|    |           |              |                                                                                                                                  |     "index_id": 25,                                                                              |
|    |           |              |                                                                                                                                  |     "filter": "((player.age==30) AND (player.name<\"zzzzz\") AND (player.name>=\"Tim Duncan\"))" |
|    |           |              |                                                                                                                                  |   },                                                                                             |
|    |           |              |                                                                                                                                  |   {                                                                                              |
|    |           |              |                                                                                                                                  |     "columnHints": [                                                                             |
|    |           |              |                                                                                                                                  |       {                                                                                          |
|    |           |              |                                                                                                                                  |         "beginValue": "42",                                                                      |
|    |           |              |                                                                                                                                  |         "endValue": "__EMPTY__",                                                                 |
|    |           |              |                                                                                                                                  |         "scanType": "PREFIX",                                                                    |
|    |           |              |                                                                                                                                  |         "column": "age"                                                                          |
|    |           |              |                                                                                                                                  |       }                                                                                          |
|    |           |              |                                                                                                                                  |     ],                                                                                           |
|    |           |              |                                                                                                                                  |     "index_id": 25,                                                                              |
|    |           |              |                                                                                                                                  |     "filter": "((player.age==42) AND (player.name<\"zzzzz\") AND (player.name>=\"Tim Duncan\"))" |
|    |           |              |                                                                                                                                  |   }                                                                                              |
|    |           |              |                                                                                                                                  | ]                                                                                                |
-----+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------
|  0 | Start     |              | ver: 0, rows: 0, execTime: 2us, totalTime: 18us                                                                                  | outputVar: [                                                                                     |
|    |           |              |                                                                                                                                  |   {                                                                                              |
|    |           |              |                                                                                                                                  |     "colNames": [],                                                                              |
|    |           |              |                                                                                                                                  |     "type": "DATASET",                                                                           |
|    |           |              |                                                                                                                                  |     "name": "__Start_0"                                                                          |
|    |           |              |                                                                                                                                  |   }                                                                                              |
|    |           |              |                                                                                                                                  | ]                                                                                                |
-----+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------
```

Reply:
Nice